PapiInstance.cpp

```diff
@@ -1,137 +1,331 @@
+		{ "TOT_CYC", PAPI_TOT_CYC },
+		{ "IPS", PAPI_IPS },
+		{ "LST_INS", PAPI_LST_INS },
+		{ "SYC_INS", PAPI_SYC_INS },
+		{ "L1_DCH", PAPI_L1_DCH },
+		{ "L2_DCH", PAPI_L2_DCH },
+		{ "L1_DCA", PAPI_L1_DCA },
+		{ "L2_DCA", PAPI_L2_DCA },
+		{ "L3_DCA", PAPI_L3_DCA },
+		{ "L1_DCR", PAPI_L1_DCR },
+		{ "L2_DCR", PAPI_L2_DCR },
+		{ "L3_DCR", PAPI_L3_DCR },
+		{ "L1_DCW", PAPI_L1_DCW },
+		{ "L2_DCW", PAPI_L2_DCW },
+		{ "L3_DCW", PAPI_L3_DCW },
+		{ "L1_ICH", PAPI_L1_ICH },
+		{ "L2_ICH", PAPI_L2_ICH },
+		{ "L3_ICH", PAPI_L3_ICH },
+		{ "L1_ICA", PAPI_L1_ICA },
+		{ "L2_ICA", PAPI_L2_ICA },
+		{ "L3_ICA", PAPI_L3_ICA },
+		{ "L1_ICR", PAPI_L1_ICR },
+		{ "L2_ICR", PAPI_L2_ICR },
+		{ "L3_ICR", PAPI_L3_ICR },
+		{ "L1_ICW", PAPI_L1_ICW },
+		{ "L2_ICW", PAPI_L2_ICW },
+		{ "L3_ICW", PAPI_L3_ICW },
+		{ "L1_TCH", PAPI_L1_TCH },
+		{ "L2_TCH", PAPI_L2_TCH },
+		{ "L3_TCH", PAPI_L3_TCH },
+		{ "L1_TCA", PAPI_L1_TCA },
+		{ "L2_TCA", PAPI_L2_TCA },
+		{ "L3_TCA", PAPI_L3_TCA },
+		{ "L1_TCR", PAPI_L1_TCR },
+		{ "L2_TCR", PAPI_L2_TCR },
+		{ "L3_TCR", PAPI_L3_TCR },
+		{ "L1_TCW", PAPI_L1_TCW },
+		{ "L2_TCW", PAPI_L2_TCW },
+		{ "L3_TCW", PAPI_L3_TCW },
+		{ "FML_INS", PAPI_FML_INS },
+		{ "FAD_INS", PAPI_FAD_INS },
+		{ "FDV_INS", PAPI_FDV_INS },
+		{ "FSQ_INS", PAPI_FSQ_INS },
+		{ "FNV_INS", PAPI_FNV_INS },
+	};
+
+	auto it = event_map.find(event_code);
+
+	if (it != event_map.end()) {
+		return it->second;
+	} else {
+		return -1;
+	}
+}
+
 void PapiW_start() {
-  printf("Starting Papi Measurement\n");
-  PapiInstance *instance = papi.create();
-  instance->addEvent(PAPI_TOT_INS);
-  instance->addEvent(PAPI_L1_ICM);
-  instance->start();
+	const char *env_codes = getenv("LIB_PAPI_EVENTS");
+
+	PapiInstance *instance = papi.create();
+
+	if (env_codes == nullptr) {
+		std::cout << "LIB_PAPI_EVENTS not defined. Measuring nothing." << std::endl;
+	} else {
+		char *codes_buffer = new char[strlen(env_codes)+1];
+		strcpy(codes_buffer, env_codes);
+
+		char delim[] = " ";
+		char *token = strtok(codes_buffer, delim);
+
+		while (token) {
+			int event = getEnvEventCode(token);
+			if (event == -1) {
+				std::cout << "[Warning] " << token
+									<< " is not a known PAPI event." << std::endl;
+			} else {
+				std::cout << "Measuring " << token << std::endl;
+				try {
+					instance->addEvent(event);
+				} catch (std::string &error) {
+					std::cout << "[Error] " << error << std::endl;
+				}
+			}
+			token = strtok(NULL, delim);
+		}
+
+		delete[] codes_buffer;
+	}
+
+	instance->start();
 }
 
 void PapiW_stopAndPrint() {
-  PapiInstance *instance = papi.create();
-  instance->stop();
-  std::cout << "\nThe measured total instructions were: "
-            << instance->getEventValue(PAPI_TOT_INS) << "\n";
-  std::cout << "Collected L1_ICM Events: "
-            << instance->getEventValue(PAPI_L1_ICM) << std::endl;
+	PapiInstance *instance = papi.create();
+	instance->stop();
+
+	const char *env_codes = getenv("LIB_PAPI_EVENTS");
```

Comment:
question: is this duplicate code?

suggestion: refactor into a function that can be used in all relevant places.